compose/decoder/decoder.go

```diff
@@ -0,0 +1,133 @@
+package decoder
+
+import (
+	"io"
+)
+
+type (
+	multiple uint
+
+	FlatReader interface {
+		Read() ([]string, error)
+	}
+
+	StructuredDecoder interface {
+		Decode(interface{}) error
+		More() bool
+	}
+
+	flatReader struct {
+		r      FlatReader
+		header []string
+		more   bool
+	}
+
+	structuredDecoder struct {
+		header []string
+		d      StructuredDecoder
+		buf    []map[string]interface{}
```

Comment:
We're storing all entries from the file in this buffer? Do I understand this correctly?

If yes -- this is not how this should be implemented, no reason to hold it in a buffer.